lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,384 @@
+{
+	int ret;
+	struct at_token tok = {0};
+
+	/* Cannot go backwards. */
+	if (!is_index_ahead(parser, index)) {
+		return -EFAULT;
+	}
+
+	do {
+		ret = at_parser_tok(parser, &tok);
+	} while (is_index_ahead(parser, index) && (ret == 0));
+
+	if (ret == 0) {
+		memcpy(token, &tok, sizeof(struct at_token));
+	}
+
+	return ret;
+}
+
+int at_parser_num_get(struct at_parser *parser, size_t index, void *value, enum at_num_type type) {
+	int ret;
+	struct at_token token = {0};
+
+	if (!parser || !value || !parser->cursor) {
+		return -EINVAL;
+	}
+
+	if (!parser->is_init) {
+		return -EPERM;
+	}
+
+	ret = at_parser_seek(parser, index, &token);
+	if (ret) {
+		return ret;
+	}
+
+	if (token.type != AT_TOKEN_TYPE_INT) {
+		return -EINVAL;
+	}
+
+	errno = 0;
+	char *next = NULL;
+	int64_t val = strtoll(token.start, &next, 10);
+
+	if (errno == ERANGE) {
+		return -ERANGE;
+	}
+
+	switch (type) {
+	case AT_NUM_INT16:
+		if (val < INT16_MIN || val > INT16_MAX) return -ERANGE;
+		*(int16_t *)value = (int16_t)val;
+		break;
+	case AT_NUM_UINT16:
+		if (val < 0 || val > UINT16_MAX) return -ERANGE;
+		*(uint16_t *)value = (uint16_t)val;
+		break;
+	case AT_NUM_INT32:
+		if (val < INT32_MIN || val > INT32_MAX) return -ERANGE;
+		*(int32_t *)value = (int32_t)val;
+		break;
+	case AT_NUM_UINT32:
+		if (val < 0 || val > UINT32_MAX) return -ERANGE;
+		*(uint32_t *)value = (uint32_t)val;
+		break;
+	case AT_NUM_INT64:
+		*(int64_t *)value = val;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int at_parser_string_get(struct at_parser *parser, size_t index, char *str, size_t *len)
+{
+	int ret;
+	struct at_token token = {0};
+
+	if (!parser || !str || !len || !parser->cursor) {
+		return -EINVAL;
+	}
+
+	if (!parser->is_init) {
+		return -EPERM;
+	}
+
+	ret = at_parser_seek(parser, index, &token);
+	if (ret) {
+		return ret;
+	}
+
+	switch (token.type) {
+	/* Acceptable types. */
+	case AT_TOKEN_TYPE_CMD_TEST:
+	case AT_TOKEN_TYPE_CMD_SET:
+	case AT_TOKEN_TYPE_CMD_READ:
+	case AT_TOKEN_TYPE_NOTIF:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_RESP:
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Check if there is enough memory for null terminator. */
+	if (*len < token.len + 1) {
+		return -ENOMEM;
+	}
+
+	memcpy(str, token.start, token.len);
+
+	/* Null-terminate the output string. */
+	str[token.len] = '\0';
+
+	/* Update the length to reflect the copied string length. */
+	*len = token.len;
+
+	return 0;
+}
+
+int at_parser_next_line_move(struct at_parser *parser)
+{
+	int ret;
+	struct at_token token = {0};
+
+	if (!parser || !parser->cursor) {
+		return -EINVAL;
+	}
+
+	if (!parser->is_init) {
+		return -EPERM;
+	}
+
+	do {
+		ret = at_parser_tok(parser, &token);
+	} while (ret == 0);
+
+	if (ret != -EAGAIN) {
+		return ret;
+	}
+
+	/* Skip CRLF. */
+	parser->cursor += 2;
+	/* Reset count. */
+	parser->count = 0;
+
+	return 0;
+}
+
+size_t at_valid_count_get(const char *at)
```

Comment:
What is this function counting? I would add that to the name and prefix the `at_parser_` part, e.g. `at_parser_tok_count_get()`

Reply:
I didn't name it with `at_parser_` prefix because it doesn't accept the `at_parser` as a parameter. This counts the number of valid parameters in the first line of the AT string. This counting function is heavily used in multiple libraries so I had to add it for some degree of compatibility with existing code.